### PR TITLE
Document Offline Signing of TAs

### PR DESCRIPTION
**Pull request by wamserma**
Documentation for https://github.com/OP-TEE/optee_os/pull/2832.

Signed-off-by: Markus S. Wamser <markus.wamser@mixed-mode.de>

### CODE REVIEW

**Comment by jbech-linaro** [comment]
Should we try to have a look at this again (and the corresponding PR in optee_os, which already have quite a few review comments)?

File: building/trusted_applications.rst [hunk 1]
```diff
@@ -29,7 +33,7 @@ To build a TA, one must provide:
     - **user_ta_header_defines.h**, a specific ANSI-C header file to define most
       of the TA properties.
 
-    - A implementation of at least the TA entry points, as extern functions:
+    - An implementation of at least the TA entry points, as extern functions:
```
Comment (jforissier) [inline building/trusted_applications.rst]: Spelling fixes are better done in a separate commit.

File: building/trusted_applications.rst [hunk 2]
```diff
@@ -299,3 +305,90 @@ sequence:
             return TEE_ERROR_NOT_SUPPORTED;
         }
     }
+
+.. _TASign:
+
+Signing of TAs
+**************
+
+All Trusted Applications are signed and verified by :ref:`optee_os` upon loading. Within the :ref:`optee_os` source is a directory ``keys``. The public parts of the keys in this directory will be compiled into the :ref:`optee_os` binary and the signature of each TA will be verified against this key(s) upon loading.
```
Comment (jforissier) [inline building/trusted_applications.rst]: Comma after `signed`? (TAs are obviously not signed "upon loading" so one should not read "signed and verified" together).
Why `keys` and `key(s)`? (plural). I'd rather be more precise: "The public part of `keys/default_ta.pem` will be [...] against this key [...]".

File: building/trusted_applications.rst [hunk 3]
```diff
@@ -299,3 +305,90 @@ sequence:
...
+
+.. warning::
+
+   :ref:`optee_os` comes with a default **private** key in its source to facilitate easy development, testing, debugging and QA.
+   Never deploy an :ref:`optee_os` binary with this key in production. Instead replace this keys as soon as possible with a public key and keep the private part of the key offline, preferably on an HSM.
```
Comment (jforissier) [inline building/trusted_applications.rst]: s/this keys/this key/

File: building/trusted_applications.rst [hunk 4]
```diff
@@ -299,3 +305,90 @@ sequence:
...
+.. code-block:: make
+
+    $(q)$(SIGN) --key $(TA_SIGN_KEY) --uuid $(user-ta-uuid) --version 0 \
+	    --in $$< --out $$@
+
+To avoid build errors when signing offline, this make script needs to be adopted. The detailed usage of ``sign.py`` is given below.
+
+Overall, offline signing is done with the following sequence of steps:
+
+0. (Preparation) Generate a 2048 bit RSA key for signing in a secure, offline environment. Extract the public key and copy it to the ``keys`` directory in the :ref:`optee_os` source tree. Adjust ``TA_SIGN_KEY`` for different file/path names. (Copy and) modify the ``link.mk`` file for the default linking step to produce a digest of the TA binary instead of the full TA.
```
Comment (jforissier) [inline building/trusted_applications.rst]: IMO if the user has to modify `link.mk`, we're doing it wrong. We'd rather want `link.mk` to support both regular ("online"?) and offline signing. In addition to making the life of users easier, it would enable automatic testing by the test suite (optee_test could select offline testing for one of its TAs for instance).

File: building/trusted_applications.rst [hunk 5]
```diff
@@ -299,3 +305,90 @@ sequence:
...
+
+3. Manually (or with an extra make target) stitch the TA together using
+
+.. code-block:: bash
+
+    sign.py stitch-ta --key $(TA_SIGN_KEY) --uuid $(user-ta-uuid) --version 0
+
+By default the UUID is taken as the base file name for all files. Different file names and paths can be set through additional options to ``sign.py``:
+
+.. code-block:: txt
```
Comment (jforissier) [inline building/trusted_applications.rst]: I recommend not duplicating the full help text for maintainability. It does not add much information here.

File: building/trusted_applications.rst (hunk 4 above)
Comment (wamserma) [reply building/trusted_applications.rst]: Yes. But documentation should also reflect the currents state of the software. So let's keep this in line with https://github.com/OP-TEE/optee_os/pull/3174 and update the doc when the `link.mk` has been updated in https://github.com/OP-TEE/optee_os.

Comment (jforissier) [reply building/trusted_applications.rst]: > update the doc when the `link.mk` has been updated in https://github.com/OP-TEE/optee_os

Fine with me.

Comment (wamserma) [reply building/trusted_applications.rst]: @jforissier I added https://github.com/OP-TEE/optee_os/issues/3275 to keep track of you comment and clear the way for this PR to be merged.

File: building/trusted_applications.rst [hunk 6]
```diff
@@ -299,3 +305,66 @@ sequence:
             return TEE_ERROR_NOT_SUPPORTED;
         }
     }
+
+.. _TASign:
+
+Signing of TAs
```
Comment (jforissier) [inline building/trusted_applications.rst]: It should be clearly stated that this applies to "REE filesystem TAs" only as described at https://optee.readthedocs.io/en/latest/architecture/trusted_applications.html#ree-filesystem-ta.

Comment (wamserma) [reply building/trusted_applications.rst]: Stated and rebased to master.

File: building/trusted_applications.rst [hunk 7]
```diff
@@ -227,6 +231,8 @@ Trusted Application properties shall be defined in a header file named
 
 Refer to :ref:`ta_properties` to understand how to configure these macros.
 
+Hint: UUIDs can be generated at https://www.guidgen.com/ or with similar tools.
```
Comment (jbech-linaro) [inline building/trusted_applications.rst]: With `.. hint::` you'll get a real "Hint notification" on the page (search for *.rst | grep hint).

File: building/trusted_applications.rst [hunk 8]
```diff
@@ -299,3 +305,66 @@ sequence:
             return TEE_ERROR_NOT_SUPPORTED;
         }
     }
+
+.. _TASign:
+
+Signing of TAs
+**************
+
+All :ref:`REE Filesystem Trusted Applications<ree_fs_ta>` need to be signed. The signature is verified by :ref:`optee_os` upon loading of the TA. Within the :ref:`optee_os` source is a directory ``keys``. The public part of ``keys/default_ta.pem`` will be compiled into the :ref:`optee_os` binary and the signature of each TA will be verified against this key upon loading.
```
Comment (jbech-linaro) [inline building/trusted_applications.rst]: This and everything below needs to be wrapped on 80 character (like it's done in all other rst files in this git).

File: building/trusted_applications.rst [hunk 9]
```diff
@@ -299,3 +305,66 @@ sequence:
...
+Currently ``keys/default_ta.pem`` must contain an RSA key.
+
+.. warning::
+
+   :ref:`optee_os` comes with a default **private** key in its source to facilitate easy development, testing, debugging and QA.
```
Comment (jbech-linaro) [inline building/trusted_applications.rst]: This is a bit of repetition of the first sentences in this document. But, maybe it's better to state it twice than having people doing mistakes.

File: building/trusted_applications.rst [hunk 10]
```diff
@@ -299,3 +305,66 @@ sequence:
...
+   Never deploy an :ref:`optee_os` binary with this key in production. Instead replace this key as soon as possible with a public key and keep the private part of the key offline, preferably on an HSM.
+
+.. note::
+
+   Currently only a single key for signing TAs is supported by :ref:`optee_os`.
+
+TAs are signed using the ``sign.py`` script referenced from ``ta/mk/ta_dev_kit.mk`` in :ref:`optee_os`. Its default behavior is to sign a compiled TA binary and attach the signature to form a complete TA for deployment. For **offline** signing, a three-step process is required: In a first step a digest of the compiled binary has to be generated, in the second step this digest is signed offline using the private key and finally in the third step the binary and its signature are stitched together into the full TA.
```
Comment (jbech-linaro) [inline building/trusted_applications.rst]: What about adding the three steps as a numbered list?
```rst
  1. A digest of the compiled binary has to be generated.
  2. The digest is signed offline using the private key.
  3. The binary and its signature are stitched together into the full TA.
```

File: building/trusted_applications.rst (hunk 8 above)
Comment (wamserma) [reply building/trusted_applications.rst]: Done.

File: building/trusted_applications.rst (hunk 10 above)
Comment (wamserma) [reply building/trusted_applications.rst]: In the context of this paragraph the visual strength of a numbered list is too strong. pulling the readers eye from the initial part of the paragraph. I prefer the smooth running text here.

Comment (jbech-linaro) [reply building/trusted_applications.rst]: Ok

**Comment by jbech-linaro** [comment]
Removed some spaces by end of lines, tweaked commit message a bit. Manually merged, closing the PR. Thanks @wamserma !
